Added a NVME FS prototype

CPTforever opened the pull request:
I made one change to Xtask that will only open the nvme.img if it already exists then truncate it, otherwise create it then truncate. 

I copied over the NVME driver code from the pager, and in the controller I detached the interrupt handler in the init_controller function in a thread pool since it seems that it doesn't work if it's solely async. I also commented out the part where it adds it to the interrupt task list. Otherwise it hangs and doesn't make any progress.

I made a quick IO handler on top of the NVME driver and then put Leo's Fat Filesystem on top of that to implement a basic file system in Twizzler.

CPTforever committed:
Added a nvme fs prototype

CPTforever committed:
Merge branch 'main' of github.com:CPTforever/twizzler into main

dbittman commented:
Looks like it's missing the submodule for the fat crate. Add it, and I'll rerun the CI.

dbittman commented on src/bin/mnemosyne/src/disk.rs:
```diff
@@ -0,0 +1,170 @@
+            }
+
+            let left = self.pos % PAGE_SIZE;
+            let right = if buf.len() - bytes_written > PAGE_SIZE {PAGE_SIZE} else {left + buf.len() - bytes_written}; // If I want to write more than the boundary of a page
```
You should run rustfmt on your code (and we should add a CI step to ensure that it's been run)

dbittman replied:
Does this if need left + buf.len() - bytes_written, similar to write, below?

dbittman commented on src/bin/mnemosyne/src/disk.rs:
```diff
@@ -0,0 +1,170 @@
+            let right = if buf.len() - bytes_written > PAGE_SIZE {PAGE_SIZE} else {left + buf.len() - bytes_written}; // If I want to write more than the boundary of a page
+            block_on(self.ctrl.read_page(lba as u64, &mut read_buffer, 0));
+
+            for p in left..right {
```
I'd suggest using [copy_from_slice](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice) and then adding the len to bytes_written.

dbittman commented on src/bin/mnemosyne/src/disk.rs:
```diff
@@ -0,0 +1,170 @@
+            }
+            
+            let temp_pos: u64 = self.pos.try_into().unwrap();
+            self.seek(SeekFrom::Start(temp_pos & !PAGE_MASK as u64)); 
```
We should be able to skip this if we are writing a whole page.

dbittman commented on src/bin/mnemosyne/src/main.rs:
```diff
@@ -0,0 +1,59 @@
+        _ => panic!("Command is invalid")
+    }
+
+    /*let str = "Top military secrets";
```
try to remove the test comments from the pr commit, but you can keep them in your local branch :)

dbittman reviewed (changes requested):
Just a few questions and comments.

CPTforever committed:
Added fat repo as just code instead of submodule

CPTforever committed:
fmted code and added optimizations

CPTforever committed:
OS Side FS STD integration

CPTforever committed:
Added std fs changes to rust

CPTforever commented:
I didn't intend to make it in the same PR but I also need to look at why the CICD failed.

But the changes made are adding 5 functions for the rust minimal runtime and reference runtime being and the corresponding functions within the std interface:
open()
read()
write()
close()
seek()

The two new functions planned will be truncate() and stat().

The way I implemented it similar to how the thread runtime was implemented with a file descriptor defined in a static map that can reference a file descriptor which contains the pos variable that points to a place in the file, and an object handle for the backed file.

The data stored within the file is the data itself and the File Metadata which holds a magic number, the size, and a direct pointer list to support extensible files. However extensible files aren't implemented at this time. 

Some limitations is that open doesn't actually create a file.

Also there is no error checking like seeing if the object is overflowing which I'm planning to rectify very soon.

gvnn3 commented on src/bin/mnemosyne/fat/layout/src/collections/array.rs:
```diff
@@ -0,0 +1,93 @@
+
+// specialization o_O
+
+// impl<const N: usize> Decode for [u8; N] {
```
Don't check in commented out code.  We use a version control system so if you need old code, you can always restore it later.

CPTforever replied:
Sorry I copied it over from Leo's repo since it's private, I didn't check it too closely, will do.

gvnn3 commented on src/bin/mnemosyne/fat/layout/layout-derive/src/lib.rs:
```diff
@@ -0,0 +1,239 @@
+                    fn framed_size(&mut self) -> Result<u64, S::Error> {
+                        Ok(#framed)
+                    }
+                    
```
There is a lot of extra white space throuhout.  Can we tighten this up?  Perhaps @dbittman has thoughts on how the code looks as well?

dbittman replied:
One thing we need to do is setup a CI step that ensures that rustfmt is happy. @CPTforever is this code formatted with rustfmt?

CPTforever replied:
I'm pretty sure I did, I ran cargo fmt in the crate's root but I'll double check.

CPTforever replied:
Ok today I found out it's not recursive, I'll commit the fix soon,

gvnn3 reviewed (changes requested):
Overall the code looks good.  I think you should add block comments to the main, visible, APIs that are used by others so they get put into the documentation by the Rust build system.

CPTforever committed:
Merge branch 'twizzler-operating-system:main' into albert-fs-std

CPTforever committed:
Fixed bug with build system

CPTforever committed:
Added error handling, minor design changes

CPTforever committed:
Merge branch 'twizzler-operating-system:main' into albert-fs-std

CPTforever committed:
Added compaction to fd allocator

CPTforever commented:
Fixed a bug with the tests so the build system should pass it now.

I've also added changes that Daniel suggested including 

Changed read and write in the runtime to accept slices instead of pointers

Defining ownedfd in std and defining a rawfd type=u32, which I implemented it by getting rid of OwnedFd and just having the file descriptor struct hold the rawfd.

Added display and error to the FsError type

Added a from impl for FsError to Io Error so that the ? operator works. I had to add a stable macro to do so, I'm not sure of another way to approach it. 

Changed the Btree + atomic counter implementation to a stable vector for counting, I also added a feature that would choose a deleted fd slot before allocating a new one.

dbittman commented on src/lib/twizzler-runtime-api/src/lib.rs:
```diff
@@ -494,8 +494,52 @@ pub struct BasicReturn {
+
+/// Runtime that implements std's FS support. Currently being implemented.
+pub trait RustFsRuntime {
+    fn open(&self, path: &CStr) -> Result<RawFd, FsError>;
```
Please document these.

dbittman reviewed (changes requested):
Just the small comments I left. I will work on fixing CI for rust submodule.

CPTforever committed:
Ran cargo fmt

CPTforever committed:
Merge branch 'albert-fs-std' of github.com:CPTforever/twizzler into albert-fs-std

CPTforever committed:
Formatted and documented the fs runtimme

CPTforever committed:
Merge branch 'albert-fs-std' into main

PandaZ3D committed:
Booting on QEMU Morello (#197)

* Add support for CHERI QEMU

* Add Morello build target

* Implement basic GICv3 support

* Run cargo fmt

* Fix interrupt initialization

* Cleanup and comments

CPTforever committed:
Merge branch 'main' of github.com:CPTforever/twizzler

CPTforever committed:
Changed the rust submodule to point to the twizzler branch

CPTforever committed:
Moved back rust submodule to previous commit

CPTforever commented:
I tried committing the latest branch in the rust submodule but I got a compilation error saying that a function in the rust runtime is missing. So I moved it back one commit.

CPTforever committed:
Fixed cargo lock because of a typo I made

gvnn3 reviewed (approved):
OK, let's get this in.
